Updated examples and README.md

i5hi opened the pull request:
- Added examples/psbt.rs

- Updated error message for examples/bip32.rs.

The previous error would print the entire path as follows:
```
not enough arguments. usage: /home/ishi/rust-bitcoin/target/debug/examples/bip32 <WIF>
```

- Updated README.md with instructions to build and run examples.

tcharding commented on examples/psbt.rs:
```diff
@@ -0,0 +1,33 @@
+
+  let args: Vec<String> = env::args().collect();
+  if args.len() < 2 {
+      eprintln!("not enough arguments. usage: cargo run --example psbt <PSBT>");
```
```suggestion
      eprintln!("Missing arguments. Usage: cargo run --example psbt <PSBT>");
```

tcharding commented on examples/psbt.rs:
```diff
@@ -0,0 +1,33 @@
+      eprintln!("not enough arguments. usage: cargo run --example psbt <PSBT>");
+      process::exit(1);
+  }
+  let decoded_psbt = match base64::decode(&args[1]){
```
```suggestion
  let decoded_psbt = match base64::decode(&args[1]) {
```
This is why we need a formatter :)

i5hi replied:
damn, clippy results:
```
error: aborting due to previous error; 113 warnings emitted
```
I'll attempt addressing this in another PR.

dr-orlovsky replied:
Now we have clippy mfixups erged so it should work for now. But also pls format this specific file with `rustfmt`

tcharding commented on examples/psbt.rs:
```diff
@@ -0,0 +1,33 @@
+  };
+
+  println!("{:#?}", &psbt_struct);
+
```
I don't rekon we need this whitespace, or that at the start of `main()`. Damn this feels like reviewing C code ...

i5hi replied:
:p sorted.

tcharding commented on Cargo.toml:
```diff
@@ -50,10 +50,14 @@ secp256k1 = { version = "0.20.0", features = [ "recovery", "rand-std" ] }
 bincode = "1.3.1"
 # We need to pin ryu (transitive dep from serde_json) to stay compatible with Rust 1.22.0
 ryu = "<1.0.5"
+base64-compat = "1.0.0"
```
Why is this here? I don't see it used anywhere in the PR?

i5hi replied:
The examples/psbt.rs file needed it to use base64::decode.

tcharding replied:
Oh nice, I didn't click that `base64-compat` had the lib name set to `base64`. Thanks

mcroad commented:
Having the same issue with rust-bitcoin/rust-miniscript#277.

To downgrade to 1.29 run `rustup default 1.29.0` and `rustc --version` to confirm.

i5hi commented:
@mcroad

i5hi commented:
@mcroad I did that. But that leads to a bunch of other errors.
Does just that command alone work for you to be able to run contrib/test.sh?

i5hi commented:
@mcroad Just saw your issue on rust-miniscript. I get the same error here but with the cc crate. Will try out a few things tomorrow and get back.

i5hi commented:
Aww man :/ Thats awkward :P Last time I attemtped this, I did run those pin commands and still got errors.

Noticed that if I ran `rustup default 1.29.0` then the pin commands and then `rustup default 1.29.0` again, it would error. I suppose `rustup default` must ONLY be run before pinning, running it both before and after (or just after) seems to reset the pin commands.

i5hi commented:
All good now. Cheers @mcroad!

i5hi commented:
Final changes to this PR are complete.

- psbt example works and is compatible with MSRV
- updated README.md with a few more pointers on testing changings with MSRV
- added psbt example to contrib/test.sh

i5hi commented:
Added --features base64 to the new psbt example in contrib/test.sh which was failing the last CI run.

dr-orlovsky commented on README.md:
````diff
@@ -61,6 +61,10 @@ please join us in
 This library should always compile with any combination of features on **Rust 1.29**.
 
+```
+rustup default 1.29.0
````
Not sure whether its good to ask user to change the default rust version - especially with the fact that it will compile with any version above that

i5hi replied:
I added this to help a contributor test with MSRV. Maybe add a note saying "To test with MSRV..." ?

dr-orlovsky replied:
For testing I would prefer something like `rustup install 1.29.0 && cargo +1.29.0 build`, which will not switch the global rust environment

i5hi replied:
Right, just `default` subcommand also  assumes 1.29.0 is already installed.

dr-orlovsky commented on README.md:
````diff
@@ -79,6 +83,8 @@ For the feature `base64` to work with 1.29.0 we also need to pin `byteorder`:
 cargo update -p byteorder --precise "1.3.4"
 ```
 
+*If you face a `Cargo.lock` parsing error, remove it before running the above commands.*
````
Or do `cargo generate-lockfile`

dr-orlovsky commented on examples/psbt.rs:
```diff
@@ -0,0 +1,34 @@
+    }
+  };
+
+  let psbt_struct: PartiallySignedTransaction = match deserialize(&decoded_psbt){
```
Nit:
```suggestion
  let psbt_struct: PartiallySignedTransaction = match deserialize(&decoded_psbt) {
```

dr-orlovsky reviewed (changes requested):
Thank you and congrats with a first-time contribution!

I left few code nits and changes to README. Unfortunately you need to rebase on top of master and remove merge commit - we do not accept from-master merges in the history. Also, it is required that each of the commits compile successfully. Sorry for inconvenience, but this is a maintainer requirement applied to all PRs in this repo.

i5hi commented:
Thanks for the support @dr-orlovsky 

Will fix this up to meet all the mentioned requirements and also address the nits you've mentioned.

tcharding commented:
Hi @i5hi, are you working on this PR still mate?

i5hi commented:
Hey. Yes.

I will re-review, update and try to close it by the end of November.

tcharding commented:
Awesome, thanks man!

i5hi commented:
oh snap, i broke it while trying to sync up.

Will make a new PR.
